### ISSUE TITLE
[Reproduction] LIBERO-Long success rate ~77.8% when training from scratch with the provided recipe (vs ~97% from released weights)

### ISSUE THREAD

**Issue by VEGA1215** [Robbyant/lingbot-va#121]
# [Reproduction] LIBERO-Long success rate ~77.8% when training from scratch with the provided recipe (vs ~97% from released weights)

## Summary

I trained a LingBot-VA video-action policy on **LIBERO-Long (`libero_10`, 10 tasks × 50 episodes = 500 episodes)** from scratch using the recipe in this repo (5000 steps, GA=10, GBS=80, lr=1e-5), but the closed-loop success rate on the full 500-episode benchmark is **77.8% (389/500)**, well below the released level.

To rule out an evaluation-side problem, I ran the **released weights** through the exact same inference config and evaluation server, and they reproduce **~97%**. This confirms the eval pipeline (policy server + LIBERO client + off-screen rendering) is correct, so the gap comes from the training reproduction, not the evaluation.

- Hardware: 8 × A800 (80GB)
- Benchmark: `libero_10`, 50 episodes / task, 500 episodes total, global seed 7
- Rendering: robosuite 1.4.1 + MuJoCo 3.1.1 (osmesa software rendering, no NVIDIA graphics stack)
- **Our from-scratch 5000-step reproduction: 77.8% (389/500), 0 episode error**
- **Released weights, same eval server: ~97%**

## Per-task results (our reproduction, 500 episodes, 0 error)

| task_id | success rate | successes / 50 |
|---:|---:|---:|
| 0 | 100% | 50 |
| 1 | 54%  | 27 |
| 2 | 80%  | 40 |
| 3 | 90%  | 45 |
| 4 | 80%  | 40 |
| 5 | 100% | 50 |
| 6 | 30%  | 15 |
| 7 | 92%  | 46 |
| 8 | 74%  | 37 |
| 9 | 78%  | 39 |
| **all** | **77.8%** | **389 / 500** |

The spread is large: tasks 0/5 are solved (100%) while task 6 collapses to 30% and task 1 to 54%. The released weights do not show this degradation on the same eval server.

## Training curves

The optimization itself converges smoothly — loss decreases monotonically and grad-norm stabilizes:

<img width="1680" height="1440" alt="Image" src="https://github.com/user-attachments/assets/ad0216f0-bb23-4011-81b7-87a4cef8447d" />

At step 5000:
- video loss (per step): **0.1189**
- action loss (per step): **0.0209**
- video loss (running mean): **0.1483**
- action loss (running mean): **0.0509**
- grad norm: **0.38**
- lr: **1.00e-05**

So the training loss looks healthy, which makes the low downstream success rate surprising.

## Reproduction configs

All values below are the actual effective config used for this run. Local paths are replaced with `/path/to/...`.

### Base / model config (`va_libero_cfg.py`)

```python
# --- diffusion / model ---
snr_shift              = 5.0        # video
action_snr_shift       = 0.05       # action
num_inference_steps    = 20         # video diffusion steps (eval)
action_num_inference_steps = 50     # action diffusion steps (eval)
guidance_scale         = 5.0        # video CFG
action_guidance_scale  = 1.0        # action CFG
frame_chunk_size       = 4
action_per_frame       = 4
attn_window            = 30
patch_size             = [1, 2, 2]
height                 = 128
width                  = 128

# --- action space ---
action_dim             = 30
used_action_channel_ids = [0, 1, 2, 3, 4, 5, 6]
inverse_used_action_channel_ids = [0, 1, 2, 3, 4, 5, 6,
                                   7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
                                   7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
                                   7, 7, 7]
action_norm_method     = "quantiles"   # per-channel q01/q99, hardcoded in repo for libero_10

# --- observation ---
obs_cam_keys = ["observation.images.agentview_rgb",
                "observation.images.eye_in_hand_rgb"]
video_exec_step = -1

# --- assets ---
wan22_pretrained_model_name_or_path = "/path/to/lingbot-va-base"
dataset_path    = "/path/to/libero-long-lerobot"
empty_emb_path  = "/path/to/libero-long-lerobot/empty_emb.pt"
```

### Train config (`va_libero_train_cfg.py`)

```python
num_steps                   = 5000
batch_size                  = 1        # per device
gradient_accumulation_steps = 10       # GBS = 1 * 10 * 8 GPUs = 80
learning_rate               = 1e-5
warmup_steps                = 10
beta1                       = 0.9
beta2                       = 0.95
weight_decay                = 0.1
cfg_prob                    = 0.1
save_interval               = 1000     # keep intermediate ckpts (1k/2k/3k/4k/5k)
load_worker                 = 16
enable_offload              = False
enable_wandb                = False
```

### Evaluation / client config

```python
# server-side inference (must match training model config)
num_inference_steps        = 20        # video diffusion steps
action_num_inference_steps = 50        # action diffusion steps
guidance_scale             = 5.0       # video CFG
action_guidance_scale      = 1.0       # action CFG
frame_chunk_size           = 4
action_per_frame           = 4
video_exec_step            = -1

# LIBERO client / rollout
benchmark          = "libero_10"
episodes_per_task  = 50                # 10 tasks -> 500 episodes
global_seed        = 7
max_steps          = 800
num_steps_wait     = 5
render_resolution  = 128
continuous_gripper = True
```

```python
# client entry (paths & host redacted)
from libero.libero.envs import OffScreenRenderEnv
from openpi_client.websocket_client_policy import WebsocketClientPolicy


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--libero-benchmark", type=str, default="libero_10")
    parser.add_argument("--port", type=int, required=True)
    parser.add_argument("--test-num", type=int, default=50)   # episodes per task
    parser.add_argument("--task_range", type=str, default="0-9")
    args = parser.parse_args()

    policy = WebsocketClientPolicy(host="0.0.0.0", port=args.port)
    # for each task in task_range: OffScreenRenderEnv(...), run test-num episodes,
    # step policy until success or max_steps (=800), count successes.


if __name__ == "__main__":
    main()
```

## Questions

Training loss converges well and our config matches the documented LIBERO recipe
(`action_snr_shift=0.05`, `used_action_channel_ids=[0..6]`, `frame_chunk_size=4`,
video/action inference steps 20/50, GBS 80, ~5k steps), yet closed-loop SR is ~20pp
below the released weights on the same eval server. Could you clarify:

1. **Global batch size**: the README mentions GBS 32/64, our run uses GBS 80. Which GBS produced the released LIBERO numbers?
2. **Steps**: the paper mentions ~4k optimizer steps, the released config / our run use 5000. Is the released checkpoint from 4k or 5k, and does a mid-training checkpoint score higher?
3. **Sequence packing / length**: the paper mentions a LIBERO sequence length of ~1e5, but we did not find sequence packing in the released `train.py`. Was packing used for the released weights?

Any pointer to which part of the training setup diverges from the released weights would be greatly appreciated. Thanks!
